File: Songhay.Dashboard/ClientApp/projects/songhay-player-video/src/lib/you-tube/models/you-tube-item.ts
import { YouTubeSnippet } from './you-tube-snippet';
import { YouTubeContentDetails } from './you-tube-content-details';

/**
 * defines YouTube API item
 * of the specified kind
 *
 * the item is the central JSON shape
 * of this API
 *
 * @export
 * @interface YouTubeItem
 */
export interface YouTubeItem {
    /**
     * entity tag to detect content changes
     *
     * @type {string}
     * @memberof YouTubeItem
     */
    etag: string;

    /**
     * unique ID for item
     *
     * @type {string}
     * @memberof YouTubeItem
     */
    id: string;

    /**
     * the kind of the item
     *
     * the kinds used so far:
     *
     * youtube#video
     * youtube#videoListResponse
     *
     * @type {string}
     * @memberof YouTubeItem
     */
    kind: string;

    /**
     * YouTube API snippet
     *
     * @type {YouTubeSnippet}
     * @memberof YouTubeItem
     */
    snippet: YouTubeSnippet;

    /**
     * YouTube item content details
     *
     * @type {YouTubeContentDetails}
     * @memberof YouTubeItem
     */
    contentDetails: YouTubeContentDetails;
}

File: Songhay.Player.YouTube/Models/YouTubeItem.fs
namespace Songhay.Player.YouTube.Models

open System
open Songhay.Modules.ProgramTypeUtility

/// <summary>
/// defines YouTube API item
/// of the specified kind
/// </summary>
/// <remarks>
/// the `item` is the central JSON shape of the YouTube API
/// </remarks>
type YouTubeItem =
    {

        /// <summary>
        /// entity tag to detect content changes
        /// </summary>
        etag: string

        /// <summary>
        /// unique ID for item
        /// </summary>
        id: string

        /// <summary>
        /// the kind of the item
        /// </summary>
        /// <remarks>
        /// the kinds used so far:
        /// - `youtube#video`
        /// - `youtube#videoListResponse`
        /// </remarks>
        kind: string

        /// <summary>
        /// YouTube API snippet
        /// </summary>
        snippet: YouTubeSnippet

        /// <summary>
        /// YouTube item content details
        /// </summary>
        contentDetails: YouTubeContentDetails
    }

    member this.getPublishedAt = this.snippet.publishedAt

    member this.tryGetDuration =
        this.contentDetails.duration |> tryParseIso8601Duration

    member this.tryGetUri : Result<Uri, exn> =
        let videoIdResult =
            match this.kind with
            | "youtube#video" -> Ok this.id
            | _ ->
                match this.snippet.resourceId with
                | Some id -> Ok id.videoId
                | _ -> Error (exn "The expected Snippet Resource ID is not here.")

        videoIdResult |> Result.map (fun videoId -> $"{YouTubeScalars.YouTubeWatchRootUri}{videoId}" |> Uri)
